[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Boyu-Zhang-UOI/pml-f2026-notebooks/blob/main/04-unsupervised/01_reading_unsupervised_results.ipynb)

# Reading an Unsupervised Result Honestly

**Sessions 15–18 · no homework depends on this, the project often does**

Unsupervised methods always return something. PCA returns components, k-means
returns clusters, t-SNE returns a picture — on structured data and on noise
alike, with no error message to tell the difference. This notebook is about the
checks that separate the two.

Run every cell; the outputs are real runs.

In [1]:
import numpy as np
from sklearn.datasets import load_digits, make_blobs
from sklearn.preprocessing import StandardScaler

rng = np.random.default_rng(0)
digits = load_digits()
X = digits.data      # raw pixels, one shared 0-16 grey scale (why unscaled: end of section 1)
y = digits.target
print("digits:", X.shape, " classes:", len(np.unique(y)),
      f" pixel values {X.min():.0f} to {X.max():.0f}")

digits: (1797, 64)  classes: 10  pixel values 0 to 16


## 1. PCA: how many components, and what they cost

`explained_variance_ratio_` is the honest summary. Read the **cumulative** curve
and pick a threshold before you look at any downstream score.

`X` holds the raw pixels, as in the Session 15 deck and reading: every column is
a grey level on the same 0–16 scale. Whether to standardize them first is a
decision, and the end of this section shows what it changes.

In [2]:
from sklearn.decomposition import PCA

pca = PCA().fit(X)
cum = np.cumsum(pca.explained_variance_ratio_)
for target in (0.50, 0.80, 0.90, 0.95, 0.99):
    k = int(np.searchsorted(cum, target)) + 1
    print(f"{target:.0%} of variance needs {k:>2} of {X.shape[1]} components")
print(f"the first two components alone keep {cum[1]:.1%}")

50% of variance needs  5 of 64 components
80% of variance needs 13 of 64 components
90% of variance needs 21 of 64 components
95% of variance needs 29 of 64 components
99% of variance needs 41 of 64 components
the first two components alone keep 28.5%


Read it as a menu, not a verdict. Five components already hold half the
variance, and sixty-four pixels drop to 29 numbers at 95%. The price climbs at
the top: going from 95% to 99%, four more points, costs twelve extra components,
and much of that tail is plausibly noise. The 50, 90, 95 and 99% rows are the
Session 15 deck's numbers (slides 20 and 21).

The other end matters too. The two components you could plot keep only 28.5%,
which is why a two-dimensional PCA picture of the digits is a smear; Session 16
starts from there.

### What the discarded components cost

A share of variance is abstract. Reconstruction puts it in the data's own units:
`transform` keeps k coordinates per image, `inverse_transform` maps them back to
64 pixels, and the gap to the original is what the dropped components held. The
cell below is the run on slide 23 of the Session 15 deck (and the reconstruction
table in the reading): same raw pixels, same values of k, so it prints the same
numbers line for line.

In [3]:
for k in (2, 8, 20, 40, 64):
    pca_k = PCA(n_components=k).fit(X)
    approx = pca_k.inverse_transform(pca_k.transform(X))
    rmse = np.sqrt(((X - approx) ** 2).mean())
    print(f"k = {k:>2}: kept "
          f"{pca_k.explained_variance_ratio_.sum():6.1%}"
          f", RMSE = {rmse:.2f} grey levels")

k =  2: kept  28.5%, RMSE = 3.66 grey levels
k =  8: kept  67.4%, RMSE = 2.47 grey levels
k = 20: kept  89.4%, RMSE = 1.41 grey levels
k = 40: kept  98.8%, RMSE = 0.47 grey levels
k = 64: kept 100.0%, RMSE = 0.00 grey levels


RMSE is the typical miss on one pixel, in grey levels on the 0–16 scale. Two
components miss by 3.66, which is the average-digit smudge on slide 22. Twenty
components bring it to 1.41, forty to 0.47, and only at `k = 64`, with nothing
discarded, does it reach zero. The k values the thresholds above picked sit
between these rows: rerun the cell with `k = 29`, the 95% choice, and see where
it lands.

`inverse_transform` hands back 64 columns, but that is a decoder re-expanding
the image for viewing. The compression is the k numbers stored per image, and
the discarded part does not come back.

### Scaling: necessary for mixed units, a choice for pixels

PCA centres the columns for you (it stores the means in `pca.mean_`, and
`inverse_transform` adds them back); that step is a rule. Scaling is not. PCA
maximises variance, so a column measured in larger units dominates the
components for no reason but its units. To see it, turn the digits into a
mixed-units table: record the busiest pixel in units a thousand times finer.

In [4]:
mixed = digits.data.copy()
loudest = int(np.argmax(mixed.var(axis=0)))   # pick a pixel that actually varies
mixed[:, loudest] = mixed[:, loudest] * 1000  # ...and measure it in different units
print(f"rescaled pixel #{loudest}")

first_unscaled = PCA(n_components=1).fit(mixed).explained_variance_ratio_[0]
first_scaled = PCA(n_components=1).fit(StandardScaler().fit_transform(mixed)).explained_variance_ratio_[0]
print(f"first component, unscaled: {first_unscaled:.4f} of variance")
print(f"first component, scaled:   {first_scaled:.4f} of variance")

rescaled pixel #42
first component, unscaled: 1.0000 of variance
first component, scaled:   0.1203 of variance


Unscaled, the first component absorbs **all** the variance: it *is* that one
rescaled pixel, and the other sixty-three have been squeezed out of the picture
by a change of units. Scaled, it is back to twelve percent. Nothing warned you;
`explained_variance_ratio_` of 1.0000 looks like a triumph. When the columns
carry different units, standardizing is necessary. Slides 10 and 11 of the
Session 15 deck show the same failure on a real table: on the raw wine
measurements PC1 is the proline column and keeps 99.81% of the variance, and
after standardizing it keeps 36%.

A detail worth copying: the pixel is chosen by `argmax(var)`. Rescaling pixel 0
instead demonstrates nothing, because in this dataset pixel 0 is a corner that
is always zero, and a thousand times zero is still zero.

Now the other direction. The real digits have no units problem: every pixel is
a grey level on the same 0–16 scale, so the differences in their variance are
information. A border pixel that is almost always blank *should* count for less
than one that carries strokes. Standardizing gives every pixel variance 1, the
nearly blank ones included:

In [5]:
pixels = digits.data
var = pixels.var(axis=0)
quiet = int(np.argmin(np.where(var > 0, var, np.inf)))   # least-used pixel that ever changes
print(f"pixel #{quiet} is non-blank in {(pixels[:, quiet] > 0).sum()} of {len(pixels)} images: "
      f"variance {var[quiet]:.4f}, against {var.max():.2f} for the busiest pixel")

for label, data in [("raw pixels         ", pixels),
                    ("standardized pixels", StandardScaler().fit_transform(pixels))]:
    cum_s = np.cumsum(PCA().fit(data).explained_variance_ratio_)
    k95 = int(np.searchsorted(cum_s, 0.95)) + 1
    print(f"{label}: top-10 variance = {cum_s[9]:.1%}, components for 95% = {k95}")

pixel #56 is non-blank in 1 of 1797 images: variance 0.0006, against 42.72 for the busiest pixel
raw pixels         : top-10 variance = 73.8%, components for 95% = 29
standardized pixels: top-10 variance = 58.9%, components for 95% = 40


Standardized, a pixel that is blank in all but one image gets the same
variance as the busiest stroke pixel, and the rarely used border pixels together
demand eleven extra components to reach 95% (40 instead of 29). Neither run is
wrong; they answer different questions. On one shared scale, standardizing is a
modeling choice: make it deliberately and be able to say why (Session 15 deck,
slides 9 to 12; the reading's *The case for leaving it alone*).

This notebook leaves the pixels raw everywhere it uses the digits, which is
what the Session 15 and 16 decks and readings do as well.

## 2. k-means always returns k clusters

Including on data with no clusters at all. Below, the same procedure runs on
three real blobs and on uniform noise; the silhouette curve is the only thing
that distinguishes them, and only if you look.

In [6]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

blobs, _ = make_blobs(n_samples=400, centers=3, cluster_std=0.6, random_state=0)
noise = rng.uniform(-8, 8, size=(400, 2))

for name, data in [("3 real blobs", blobs), ("uniform noise", noise)]:
    line = []
    for k in range(2, 7):
        km = KMeans(n_clusters=k, n_init=10, random_state=0).fit(data)
        line.append(f"k={k}: sil={silhouette_score(data, km.labels_):.2f}")
    print(f"{name:<14} " + "  ".join(line))

3 real blobs   k=2: sil=0.55  k=3: sil=0.66  k=4: sil=0.54  k=5: sil=0.42  k=6: sil=0.33
uniform noise  k=2: sil=0.37  k=3: sil=0.38  k=4: sil=0.41  k=5: sil=0.39  k=6: sil=0.40


The blobs peak sharply at `k=3` (0.66) and fall away to 0.33 by `k=6`. The noise
curve is **flat** — it drifts between 0.37 and 0.41 with no k standing clear of
its neighbours, which is the signature of "there is no cluster structure here".

Note that the noise curve has a nominal maximum, at `k=4`. Reporting *that*
number alone would look like a result. It is the shape of the curve, not its
argmax, that carries the information.

Report the curve, not the single number at your chosen k.

## 3. DBSCAN finds shapes k-means cannot — and refuses to answer

k-means partitions space with straight boundaries, so it cannot represent a
crescent. DBSCAN can, and it also labels points `-1` when they belong to no
dense region, which is the honesty k-means lacks.

In [7]:
from sklearn.cluster import DBSCAN
from sklearn.datasets import make_moons
from sklearn.metrics import adjusted_rand_score

moons, moon_labels = make_moons(n_samples=400, noise=0.06, random_state=0)

km = KMeans(n_clusters=2, n_init=10, random_state=0).fit(moons)
db = DBSCAN(eps=0.25, min_samples=5).fit(moons)

print(f"k-means  ARI={adjusted_rand_score(moon_labels, km.labels_):.3f}")
print(f"DBSCAN   ARI={adjusted_rand_score(moon_labels, db.labels_):.3f}  "
      f"noise points: {(db.labels_ == -1).sum()}")

k-means  ARI=0.274
DBSCAN   ARI=1.000  noise points: 0


`eps` is the parameter that matters and it has no default worth trusting. Too
small and everything is noise; too large and everything is one cluster:

In [8]:
for eps in (0.10, 0.18, 0.25, 0.60):
    labels = DBSCAN(eps=eps, min_samples=5).fit_predict(moons)
    n_clusters = len(set(labels) - {-1})
    print(f"eps={eps:<5} clusters={n_clusters}  noise={(labels == -1).sum():>3}  "
          f"ARI={adjusted_rand_score(moon_labels, labels):.3f}")

eps=0.1   clusters=4  noise= 10  ARI=0.731
eps=0.18  clusters=2  noise=  0  ARI=1.000
eps=0.25  clusters=2  noise=  0  ARI=1.000
eps=0.6   clusters=1  noise=  0  ARI=0.000


## 4. t-SNE: what the picture is allowed to tell you

t-SNE preserves neighbourhoods and discards global geometry. Two consequences
that are easy to state and easy to forget under a pretty plot: **cluster sizes
mean nothing**, and **distances between clusters mean nothing**.

The check that costs one line is to run it again with another seed. The input
is the first 500 digits as raw pixels: the same `X` as section 1, and the same
preprocessing as Session 16's deck and reading. Nothing here forces
standardizing; t-SNE reads distances between images, and the pixels already
share one grey scale.

The seed only matters when the starting layout is random. With `init="pca"`,
the default, the scikit-learn this notebook was built with (1.9) starts every
seed from the same PCA layout, and two seeds print the same map twice: a stable
picture, not a checked one. The runs below start at random, so the seed has
something to change. The labels are used only afterwards, to find where the 0s
landed; t-SNE never sees them.

In [9]:
from sklearn.manifold import TSNE
from sklearn.neighbors import NearestNeighbors

sub, sub_y = X[:500], y[:500]

def neighbours(emb, k=10):
    # each point's k nearest neighbours in the picture, itself excluded
    nn = NearestNeighbors(n_neighbors=k + 1).fit(emb)
    return nn.kneighbors(emb, return_distance=False)[:, 1:]

runs = {s: TSNE(n_components=2, perplexity=30, init="random",
                random_state=s).fit_transform(sub) for s in (0, 1, 2)}
first = neighbours(runs[0])
for s, emb in runs.items():
    zeros = emb[sub_y == 0].mean(axis=0)
    shared = np.mean([len(set(a) & set(b)) / 10 for a, b in zip(first, neighbours(emb))])
    print(f"seed {s}: the 0s land at {np.round(zeros, 1)}, "
          f"10 nearest neighbours shared with seed 0: {shared:.2f}")

seed 0: the 0s land at [ -2.  -32.9], 10 nearest neighbours shared with seed 0: 1.00
seed 1: the 0s land at [  4.6 -34.7], 10 nearest neighbours shared with seed 0: 0.97
seed 2: the 0s land at [17.9 21.6], 10 nearest neighbours shared with seed 0: 0.98


Where a group lands moves with the seed: seed 1 happens to put the 0s near
where seed 0 did, and seed 2 sends them to the opposite side of the map. Which
points sit together barely moves: on average a point keeps 97–98% of the ten
nearest neighbours it had under seed 0. Only the second is a property of the
data, so runs are not comparable point-for-point, and the neighbourhoods are
what should survive.

If a structure appears under one seed and not another, it is a property of the
optimiser, not of your data. Say in the caption which seed produced the figure
you printed.

---

## Using this in the project

- **PCA** — report the cumulative variance curve and the reconstruction error at
  the k you chose, not just the count. Standardize when the columns carry
  different units; on a shared scale such as pixels it is a choice, so say which
  you made and why.
- **k-means** — report silhouette across a range of k. A flat curve means no
  cluster structure, and that is a finding worth writing down rather than hiding.
- **DBSCAN** — report `eps` and the noise-point count; a run that labels 60% of
  the data as noise is not a clustering.
- **t-SNE** — two seeds from a random start, and a caption naming the one you
  show.

## Where to go next

- **Reading, Sessions 15–18** — the mechanics behind each of these checks.
- **Session 15's reading** on when standardizing helps PCA and when it hurts.
- **Session 16's reading** on why t-SNE distances are decoration.